### PR TITLE
Create an extensions tutorial

### PR DESCRIPTION
**Pull request by WilliamJamieson**

### CODE REVIEW

File: notebooks/06-Custom-Extensions.ipynb [hunk 1]
```diff
@@ -0,0 +1,577 @@
+{
```
Comment (zacharyburnett) [inline notebooks/06-Custom-Extensions.ipynb:1]: <p>should references to other chapters have links?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/asdf-format/scipy2022tutorial/pull/8/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

File: notebooks/06-Custom-Extensions.ipynb [hunk 2]
```diff
@@ -0,0 +1,577 @@
...
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Introduction to Writing an Extension\n",
+    "\n",
+    "Our ultimate goal is to create a `Converter` class which converts from an `Ellipse` object\n",
+    "to an ASDF file and from an ASDF file back into an `Ellipse` object, such that an `Ellipse`\n",
+    "object \"round-trips\".\n",
+    "\n",
+    "Formally `Converter` interface defines a mapping between tagged objects in the `ASDF` tree\n",
```
Comment (zacharyburnett) [inline notebooks/06-Custom-Extensions.ipynb:65]:
```suggestion
    "Formally, the `Converter` interface defines a mapping between tagged objects in the `ASDF` tree\n",
```

File: notebooks/06-Custom-Extensions.ipynb [hunk 3]
```diff
@@ -0,0 +1,577 @@
...
+    "and their corresponding Python objects. Typically, this means one converter per tag/object\n",
+    "pair; however, there is support many-to-one and many-to-many mappings for a single `Converter`\n",
```
Comment (zacharyburnett) [inline notebooks/06-Custom-Extensions.ipynb:67]:
```suggestion
    "pair; however, a single `Converter` can also support many-to-one and many-to-many mappings \n",
```

File: notebooks/06-Custom-Extensions.ipynb [hunk 4]
```diff
@@ -0,0 +1,577 @@
...
+    "\n",
+    "All converters should be constructed as subclasses of the abstract type `asdf.extension.Converter`,\n",
+    "which requires that you define two methods:\n",
+    "\n",
+    "1. `to_yaml_tree`: which converts a Python object into an ASDF tree.\n",
+    "2. `from_yaml_tree`: which converts an ASDF tree into a python object.\n",
+    "\n",
+    "Note that these methods can account for the type/tag of the objects attempting to be converted.\n",
+    "\n",
+    "Moreover your converter also needs to define to variables:\n",
```
Comment (zacharyburnett) [inline notebooks/06-Custom-Extensions.ipynb:252]:
```suggestion
    "Moreover your converter also needs to define the following two variables:\n",
```

File: notebooks/06-Custom-Extensions.ipynb [hunk 5]
```diff
@@ -0,0 +1,577 @@
...
+    "\n",
+    "1. `tags`: A list of the ASDF tags that this converter will be used by ASDF to read.\n",
```
Comment (zacharyburnett) [inline notebooks/06-Custom-Extensions.ipynb:254]:
```suggestion
    "1. `tags`: A list of tags that this converter will use when reading ASDF.\n",
```

File: notebooks/06-Custom-Extensions.ipynb [hunk 6]
```diff
@@ -0,0 +1,577 @@
...
+    "2. `types`: A list of the Python (object) types that this converter will be used by ASDF write.\n",
```
Comment (zacharyburnett) [inline notebooks/06-Custom-Extensions.ipynb:255]:
```suggestion
    "2. `types`: A list of Python (object) types that this converter will use when writing ASDF.\n",
```

File: notebooks/06-Custom-Extensions.ipynb (hunk 1 above)
Comment (WilliamJamieson) [reply notebooks/06-Custom-Extensions.ipynb:1]: <p>How do you create links to other tutorials?</p>

File: notebooks/06-Custom-Extensions.ipynb [hunk 7]
```diff
@@ -0,0 +1,577 @@
...
+    "\n",
+    "    def from_yaml_tree(self, node, tag, ctx):\n",
+    "        return Ellipse(semi_major=node[\"semi_major\"], semi_minor=node[\"semi_minor\"])"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Note that one will normally defer the import of the object to be created to inside `from_yaml_tree`\n",
```
Comment (zacharyburnett) [inline notebooks/06-Custom-Extensions.ipynb:288]:
```suggestion
    "Note that, for performance of the entry points, one will normally defer the `import` of the object to be created\n",
```

File: notebooks/06-Custom-Extensions.ipynb [hunk 8]
```diff
@@ -0,0 +1,577 @@
...
+    "for performance of the entry points."
```
Comment (zacharyburnett) [inline notebooks/06-Custom-Extensions.ipynb:289]:
```suggestion
    "until `from_yaml_tree` is actually called."
```

File: notebooks/06-Custom-Extensions.ipynb [hunk 9]
```diff
@@ -0,0 +1,577 @@
...
+   "metadata": {},
+   "source": [
+    "\n",
+    "ASDF provides the `asdf.resource.DirectoryResourceMapping` object to crawl resource directories.\n",
+    "It allows us to turn these directory structures into objects which can subsequently be added to\n",
+    "ASDF using the entry points.\n",
+    "\n",
+    "These objects require two input parameters:\n",
+    "1. A path to the root directory which contains the resources to be added.\n",
+    "2. The prefix that will be used as a prefix together with the file names to generate the URI\n",
```
Comment (zacharyburnett) [inline notebooks/06-Custom-Extensions.ipynb:444]:
```suggestion
    "2. The prefix that will be used together with the file names to generate the URI\n",
```

**Review by zacharyburnett — COMMENTED** [review]
minor fixes for clarity

File: notebooks/06-Custom-Extensions.ipynb (hunk 7 above)
Comment (zacharyburnett) [reply notebooks/06-Custom-Extensions.ipynb]: these performance improvements should be even more pronounced when [PEP690](https://peps.python.org/pep-0690/) is implemented

**Pull request merged**